### PR TITLE
[WIP] Add maintainership

### PR DESCRIPTION
**Pull request by ErikSchierboom**

### CODE REVIEW

File: app/models/track/maintainership.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+class Track::Maintainership < ApplicationRecord
+  belongs_to :track
+  belongs_to :maintainer,
+    class_name: "User",
+    foreign_key: :user_id,
+    inverse_of: :maintainerships
+
+  enum maintainer_type: { regular: 0, polyglot: 1 }, _suffix: :maintainer
```
Comment (ErikSchierboom) [inline app/models/track/maintainership.rb]: Let me know what you think about this name. I've added the suffix here as `:regular_maintainer` felt better than `:maintainer_regular`.

Comment (ErikSchierboom) [reply app/models/track/maintainership.rb]: I've toyed with having `track` and `polyglot` values for the type, as well as making `polyglot` a `bool`. The current setup requires us to add any polyglot maintainers to each track, but I think that's fine.

File: app/commands/track/maintainership/create.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
+class Track::Maintainership
+  class Create
```
Comment (ErikSchierboom) [inline app/commands/track/maintainership/create.rb]: I've only created this command, and have not done anything with updating/removing maintainership. Want me to add those?

File: app/models/track/maintainership.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
+class Track::Maintainership < ApplicationRecord
+  belongs_to :maintainer,
+    class_name: "User",
+    foreign_key: :user_id,
+    inverse_of: :maintainerships
+
+  belongs_to :component, polymorphic: true
+
+  enum maintainer_type: { track: 0, review: 1 }, _suffix: :maintainer
```
Comment (iHiD) [inline app/models/track/maintainership.rb]:
```suggestion
  enum maintainer_type: { track: 0, reviewer: 1 }, _suffix: :maintainer
```

File: app/models/track/maintainership.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  enum maintainer_level: { apprentice: 0, regular: 1, senior: 2 }, _suffix: :maintainer
+  enum component_type: { Track: 0 }, _prefix: true
```
Comment (iHiD) [inline app/models/track/maintainership.rb]:
```suggestion
  enum component_type: { track: 0 }, _prefix: true
```

File: app/models/track/maintainership.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  def maintainer_level
+    super.to_sym
+  end
+
+  def component_type
+    super.to_sym
+  end
+
+  def track
+    return nil unless component_type == :Track
```
Comment (iHiD) [inline app/models/track/maintainership.rb]:
```suggestion
    return nil unless component_type == :track
```

File: app/models/track/maintainership.rb [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    super.to_sym
+  end
+
+  def maintainer_level
+    super.to_sym
+  end
+
+  def component_type
+    super.to_sym
+  end
```
Comment (iHiD) [inline app/models/track/maintainership.rb]: You could meta-program these:
```ruby
%i{maintainer_type maintainer_level component_type}.each do |meth|
define_method meth { super.to_sym}
end
```

File: app/models/user.rb [hunk 7]
```diff
@@ -42,6 +42,9 @@ class User < ApplicationRecord
   has_many :track_mentorships, dependent: :destroy
   has_many :mentored_tracks, through: :track_mentorships, source: :track
 
+  has_many :maintainerships, class_name: "Track::Maintainership", dependent: :destroy
+  has_many :maintained_tracks, through: :maintainerships, source: :track
```
Comment (iHiD) [inline app/models/user.rb]: I don't **think** you can have a source that's not an AR relationship. You could make track into an AR relationship though using an `optional` `belongs_to` with a `where` in maintainership.

**Pull request closed without merging**